api/config/v1alpha1/validation/envoyproxy.go

```diff
@@ -41,17 +41,47 @@ func validateEnvoyProxySpec(spec *egcfgv1a1.EnvoyProxySpec) error {
 	if spec == nil {
 		errs = append(errs, errors.New("spec is nil"))
 	}
-	if spec != nil && spec.Provider != nil && spec.Provider.Type != egcfgv1a1.ProviderTypeKubernetes {
-		errs = append(errs, fmt.Errorf("unsupported provider type %v", spec.Provider.Type))
+
+	// validate provider
+	validateProviderErrs := validateProvider(spec)
+	if len(validateProviderErrs) != 0 {
+		errs = append(errs, validateProviderErrs...)
 	}
-	if spec.Bootstrap != nil {
+
+	// validate bootstrap
+	if spec != nil && spec.Bootstrap != nil {
 		if err := validateBootstrap(spec.Bootstrap); err != nil {
 			errs = append(errs, err)
 		}
 	}
 	return utilerrors.NewAggregate(errs)
 }
 
+func validateProvider(spec *egcfgv1a1.EnvoyProxySpec) []error {
+	var errs []error
+	if spec != nil && spec.Provider != nil {
+		if spec.Provider.Type != egcfgv1a1.ProviderTypeKubernetes {
+			errs = append(errs, fmt.Errorf("unsupported provider type %v", spec.Provider.Type))
+		}
+		validateServiceTypeErrs := validateServiceType(spec)
+		if len(validateServiceTypeErrs) != 0 {
+			errs = append(errs, validateServiceTypeErrs...)
+		}
+	}
+	return errs
+}
+
+func validateServiceType(spec *egcfgv1a1.EnvoyProxySpec) []error {
+	var errs []error
+	if spec.Provider.Kubernetes != nil && spec.Provider.Kubernetes.EnvoyService != nil {
+		serviceType := spec.Provider.Kubernetes.EnvoyService.Type
+		if *serviceType != egcfgv1a1.ServiceTypeLoadBalancer && *serviceType != egcfgv1a1.ServiceTypeClusterIP {
```

Comment:
trying to apply following configuration:
```
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: example-config
spec:
  provider:
    type: Kubernetes
    kubernetes:
      envoyService:
        type: NodePort
```
you will get following message: `The EnvoyProxy "example-config" is invalid: spec.provider.kubernetes.envoyService.type: Unsupported value: "NodePort": supported values: "LoadBalancer", "ClusterIP"`
IMO, this check is redundant。

Reply:
so you mean the API-server has checked this?

Reply:
But I think the validation here is necessary for other modules inside eg

Reply:
> so you mean the API-server has checked this?

yes

> But I think the validation here is necessary for other modules inside eg, like test and controller etc

EG watch it through client-go?

Reply:
yes this check is not needed since the kube builder annotation bakes it into the CRD that is used by the API server to validate the data

Reply:
but we probably need a `nil` check since its a pointer, right ?

Reply:
> but we probably need a `nil` check since its a pointer, right ?

nil will use the default, but i think double check is necessary? @arkodg